File: i2f-jdk/i2f-swl/web/i2f-swl/src/i2f-turbo-web/i2f-swl/core/core/key/SwlKeyExchanger.js
import SwlExchanger from "../exchanger/SwlExchanger";
import SwlStorageKeyManager from "@/i2f-turbo-web/i2f-swl/core/core/key/impl/SwlStorageKeyManager";

/**
 * @param keyManager {SwlKeyManager}
 * @return {SwlKeyExchanger}
 * @extends {SwlExchanger}
 * @constructor {SwlKeyExchanger}
 */
function SwlKeyExchanger(keyManager=new SwlStorageKeyManager()) {
    /**
     *
     * @type {SwlKeyManager}
     */
    this.keyManager=keyManager
}


// 继承
SwlKeyExchanger.prototype = Object.create(SwlExchanger.prototype)
SwlKeyExchanger.prototype.constructor = SwlKeyExchanger


/**
 * @return {AsymKeyPair}
 */
SwlKeyExchanger.prototype.getSelfKeyPair = function () {
    this.keyManager.getSelfKeyPair()
}


/**
 *
 * @return {String}
 */
SwlKeyExchanger.prototype.getSelfPublicKey=function() {
    let keyPair = this.getSelfKeyPair();
    return keyPair.getPublicKey();
}

/**
 *
 * @param publicKey {String}
 * @return {String}
 */
SwlKeyExchanger.prototype.calcKeySign=function(publicKey) {
    return this.messageDigester.digest(publicKey);
}

/**
 *
 * @param selfAsymSign {String}
 * @param selfKeyPair {AsymKeyPair}
 * @return {void}
 */
SwlKeyExchanger.prototype.setSelfKeyPair=function(selfAsymSign, selfKeyPair) {
    this.keyManager.setSelfKeyPair(selfAsymSign, selfKeyPair);
    this.keyManager.setDefaultSelfAsymSign(selfAsymSign);
}

/**
 *
 * @param selfAsymSign {String}
 * @return {String|null}
 */
SwlKeyExchanger.prototype.getSelfPrivateKey=function( selfAsymSign) {
    let keyPair = this.keyManager.getSelfKeyPair(selfAsymSign);
    if (!keyPair) {
        return null;
    }
    return keyPair.getPrivateKey();
}

/**
 *
 * @param otherAsymSign {String}
 * @return {String|null}
 */
SwlKeyExchanger.prototype.getOtherPublicKey=function(otherAsymSign) {
    return this.keyManager.getOtherPublicKey(otherAsymSign);
}

/**
 *
 * @return {String|null}
 */
SwlKeyExchanger.prototype.getOtherPublicKeyDefault=function() {
    return this.keyManager.getOtherPublicKey();
}

/**
 *
 * @param otherAsymSign {String}
 * @param publicKey {String}
 * @return {void}
 */
SwlKeyExchanger.prototype.setOtherPublicKey=function(otherAsymSign, publicKey) {
    this.keyManager.setOtherPublicKey(otherAsymSign, publicKey);
    this.keyManager.setDefaultOtherAsymSign(otherAsymSign);
}

/**
 *
 * @param otherPublicKey {String}
 * @return {void}
 */
SwlKeyExchanger.prototype.acceptOtherPublicKey=function(otherPublicKey) {
    let otherAsymSign = this.messageDigester.digest(otherPublicKey);
    this.setOtherPublicKey(otherAsymSign, otherPublicKey);
}

/**
 *
 * @return {String|null}
 */
SwlKeyExchanger.prototype.getSelfSwapKey=function() {
    let selfPublicKey = this.getSelfPublicKey();
    return this.obfuscateEncode(selfPublicKey);
}

/**
 *
 * @param otherSwapKey {String}
 * @return {void}
 */
SwlKeyExchanger.prototype.acceptOtherSwapKey=function(otherSwapKey) {
    let otherPublicKey = this.obfuscateDecode(otherSwapKey);
    this.acceptOtherPublicKey(otherPublicKey);
}

/**
 *
 * @return {AsymKeyPair}
 */
SwlKeyExchanger.prototype.resetSelfKeyPair=function() {
    let asymmetricEncryptor = this.asymmetricEncryptorSupplier.get();
    let asymKeyPair = asymmetricEncryptor.generateKeyPair();
    let selfAsymSign = this.messageDigester.digest(asymKeyPair.getPublicKey());
    this.setSelfKeyPair(selfAsymSign, asymKeyPair);
    return asymKeyPair;
}

/**
 *
 * @param remotePublicKey {String}
 * @param parts {string[]}
 * @param attaches {string[]|null}
 * @return {SwlData}
 */
SwlKeyExchanger.prototype.sendByKey=function(remotePublicKey,  parts,  attaches=[]) {
    let keyPair = this.getSelfKeyPair();
    return this.sendByRaw(remotePublicKey, this.calcKeySign(remotePublicKey),
        keyPair.getPrivateKey(), this.calcKeySign(keyPair.getPublicKey()),
        parts, attaches);
}

/**
 *
 * @param clientId {String}
 * @param request {SwlData}
 * @return {SwlData}
 */
SwlKeyExchanger.prototype.receiveByKey=function(clientId, request) {
    let selfAsymSign = request.header.remoteAsymSign;
    let otherAsymSign = request.header.localAsymSign;
    let otherPublicKey = this.getOtherPublicKey(otherAsymSign);
    let selfPrivateKey = this.getSelfPrivateKey(selfAsymSign);
    return this.receiveByRaw(clientId, request, otherPublicKey, selfPrivateKey);
}
/**
 *
 * @param parts {string[]}
 * @param attaches {string[]|null}
 * @return {SwlData}
 */
 SwlKeyExchanger.prototype.sendDefaultByKey=function(parts, attaches=[]) {
    let otherPublicKey = this.getOtherPublicKeyDefault();
    return this.sendByKey(otherPublicKey, parts, attaches);
}

/**
 *
 * @param remoteAsymSign {String}
 * @param parts {string[]}
 * @param attaches {string[]|null}
 * @return {SwlData}
 */
 SwlKeyExchanger.prototype.responseByKey=function( remoteAsymSign,  parts, attaches=[]) {
    let otherPublicKey = this.getOtherPublicKey(remoteAsymSign);
    return this.sendByKey(otherPublicKey, parts, attaches);
}


export default SwlKeyExchanger

File: i2f-jdk/i2f-swl/src/main/java/i2f/swl/core/key/SwlKeyExchanger.java
package i2f.swl.core.key;

import i2f.crypto.std.encrypt.asymmetric.key.AsymKeyPair;
import i2f.swl.core.exchanger.SwlExchanger;
import i2f.swl.core.key.impl.SwlLocalFileKeyManager;
import i2f.swl.data.SwlData;
import i2f.swl.std.ISwlAsymmetricEncryptor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.List;

/**
 * @author Ice2Faith
 * @date 2024/7/9 20:22
 * @desc
 */
@Data
@NoArgsConstructor
public class SwlKeyExchanger extends SwlExchanger {

    protected SwlKeyManager keyManager = new SwlLocalFileKeyManager();

    public AsymKeyPair getSelfKeyPair() {
        return keyManager.getDefaultSelfKeyPair();
    }

    public String getSelfPublicKey() {
        AsymKeyPair keyPair = getSelfKeyPair();
        return keyPair.getPublicKey();
    }

    public String calcKeySign(String publicKey) {
        return messageDigester.digest(publicKey);
    }

    public void setSelfKeyPair(String selfAsymSign, AsymKeyPair selfKeyPair) {
        keyManager.setSelfKeyPair(selfAsymSign, selfKeyPair);
        keyManager.setDefaultSelfAsymSign(selfAsymSign);
    }

    public String getSelfPrivateKey(String selfAsymSign) {
        AsymKeyPair keyPair = keyManager.getSelfKeyPair(selfAsymSign);
        if (keyPair == null) {
            return null;
        }
        return keyPair.getPrivateKey();
    }

    public String getOtherPublicKey(String otherAsymSign) {
        return keyManager.getOtherPublicKey(otherAsymSign);
    }

    public String getOtherPublicKeyDefault() {
        return keyManager.getDefaultOtherPublicKey();
    }

    public void setOtherPublicKey(String otherAsymSign, String publicKey) {
        keyManager.setOtherPublicKey(otherAsymSign, publicKey);
        keyManager.setDefaultOtherAsymSign(otherAsymSign);
    }

    public void acceptOtherPublicKey(String otherPublicKey) {
        String otherAsymSign = messageDigester.digest(otherPublicKey);
        setOtherPublicKey(otherAsymSign, otherPublicKey);
    }

    public String getSelfSwapKey() {
        String selfPublicKey = getSelfPublicKey();
        return obfuscateEncode(selfPublicKey);
    }

    public void acceptOtherSwapKey(String otherSwapKey) {
        String otherPublicKey = obfuscateDecode(otherSwapKey);
        acceptOtherPublicKey(otherPublicKey);
    }

    public AsymKeyPair resetSelfKeyPair() {
        ISwlAsymmetricEncryptor asymmetricEncryptor = requireAsymmetricEncryptor();
        AsymKeyPair asymKeyPair = asymmetricEncryptor.generateKeyPair();
        releaseAsymmetricEncryptor(asymmetricEncryptor);
        String selfAsymSign = messageDigester.digest(asymKeyPair.getPublicKey());
        setSelfKeyPair(selfAsymSign, asymKeyPair);
        return asymKeyPair;
    }

    public SwlData sendByKey(String remotePublicKey, List<String> parts) {
        return sendByKey(remotePublicKey, parts, null);
    }

    public SwlData sendByKey(String remotePublicKey, List<String> parts, List<String> attaches) {
        AsymKeyPair keyPair = getSelfKeyPair();
        return sendByRaw(remotePublicKey, calcKeySign(remotePublicKey),
                keyPair.getPrivateKey(), calcKeySign(keyPair.getPublicKey()),
                parts, attaches);
    }

    public SwlData receiveByKey(String clientId, SwlData request) {
        String selfAsymSign = request.getHeader().getRemoteAsymSign();
        String otherAsymSign = request.getHeader().getLocalAsymSign();
        String otherPublicKey = getOtherPublicKey(otherAsymSign);
        String selfPrivateKey = getSelfPrivateKey(selfAsymSign);
        return receiveByRaw(clientId, request, otherPublicKey, selfPrivateKey);
    }

    public SwlData sendDefaultByKey(List<String> parts, List<String> attaches) {
        String otherPublicKey = getOtherPublicKeyDefault();
        return sendByKey(otherPublicKey, parts, attaches);
    }

    public SwlData sendDefaultByKey(List<String> parts) {
        String otherPublicKey = getOtherPublicKeyDefault();
        return sendByKey(otherPublicKey, parts, null);
    }

    public SwlData responseByKey(String remoteAsymSign, List<String> parts, List<String> attaches) {
        String otherPublicKey = getOtherPublicKey(remoteAsymSign);
        return sendByKey(otherPublicKey, parts, attaches);
    }

    public SwlData responseByKey(String remoteAsymSign, List<String> parts) {
        String otherPublicKey = getOtherPublicKey(remoteAsymSign);
        return sendByKey(otherPublicKey, parts, null);
    }
}
